File: src/java/sort/BubbleSort.java
package sort;

/*
 * Sample program for Bubble Sort algorithm
 *
 * https://www.geeksforgeeks.org/bubble-sort/
 *
 * Bubble Sort repeatedly swaps the adjacent elements if they are in wrong order. The algorithm stops when it goes through one pass
 * without any swaps.
 *
 * Worst and Average Case Time Complexity: O(n*n). Worst case occurs when array is reverse sorted.
 * In place algorithm
 * Stable algorithm
 */

public class BubbleSort {
    public static void main(String args[]) {
        int[] arr = {20, 35, -15, 7, 55, 1, -22};

        System.out.print("Unsorted array: ");
        printArray(arr);

        // lastUnsortedIndex goes right to left while i goes left to right based on the lastUnsortedIndex
        for (int lastUnsortedIndex = arr.length - 1; lastUnsortedIndex > 0; lastUnsortedIndex--) {
            for (int i = 0; i < lastUnsortedIndex; i++) {
                if (arr[i] > arr[i + 1]) {
                    swap(arr, i, i + 1);
                }
            }
        }

        System.out.print("Sorted array: ");
        printArray(arr);
    }

    public static void swap(int[] arr, int i, int j) {
        if (i == j) {
            return;
        }
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: src/c++/sort/BubbleSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void sort(vector<int>& arr);

/*
 * BUBBLE SORT ALGORITHM
 * 
 * Repeatedly swaps the adjacent elements if they are in wrong order.
 * It stops when it goes through one pass without any swaps.
 *
 * O(N^2), comparison/stable, inplace
 * 
 */

void sort(vector<int>& arr) {
    bool swap = true;
    while (swap) {
        swap = false;
        for (int i = 0; i < arr.size() - 1; i++) {
            if (arr[i + 1] < arr[i]) {
                int temp = arr[i + 1];
                arr[i + 1] = arr[i];
                arr[i] = temp;
                swap = true;
            }
        }
    }
}

int main() {
    vector<int> arr = {6, 3, -1, 0, 4, 2, 1, 3, 5}; // -1 0 1 2 3 3 4 5 6
    sort(arr);
    for (int i : arr) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}